### PR TITLE
Implement core logic for TOTP support

### PR DESCRIPTION
**Pull request by zoldar**
### Changes

This PR implements core logic for adding 2FA support using TOTP.

The most important part is `Plausible.Auth.TOTP` exposing API for all things related to TOTP setup, management and validation. It's moduledoc is a good starting point for review. That aside, the PR can be reviewed commit by commit.

Before merging, `TOTP_VAULT_KEY` must be set in infrastructure config for all relevant envs.

The migration should be extracted and run before merging this PR.

With all that in mind, the PR should be safe to merge and deploy without affecting any existing functionality.

The follow-up work will include email notification on enabling and disabling TOTP and finally the UI.

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [ ] Entry has been added to changelog
- [x] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [x] This change does not need a documentation update

### Dark mode
- [ ] The UI has been tested both in dark and light mode
- [x] This PR does not change the UI

### CODE REVIEW

**Review by vinibrsl — APPROVED** [review]
✨

File: lib/plausible/auth/totp.ex [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  TOTP auth context
+
+  Handles all the aspects of TOTP setup, management and validation for users.
+
+  ## Setup
+
+  TOTP setup is started with `initiate/1`. At this stage, a random secret
+  binary is generated for user and stored under `User.totp_secret`. The secret
+  is additionally encrypted while stored in the database using `Cloak`. The
+  vault for stafe storage is configured in `Plausible.Auth.TOTP.Vault` via
```
Comment (aerosol) [inline lib/plausible/auth/totp.ex:12]:
```suggestion
  vault for safe storage is configured in `Plausible.Auth.TOTP.Vault` via
```

File: lib/plausible/auth/totp.ex [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  a dedicated `Ecto` type defined in `Plausible.Auth.TOTP.EncryptedBinary`.
+  The function returns updated user along with TOTP URI and a readable form
+  of secret. Both are meant for exposing in user's setup screen. The URI
```
Comment (aerosol) [inline lib/plausible/auth/totp.ex:15]:
```suggestion
  of secret. Both - the URI and readable secret - are meant for exposure in the user's setup screen. The URI
```

File: lib/plausible/auth/totp.ex (hunk 1 above)
Comment (aerosol) [reply lib/plausible/auth/totp.ex:12]: or state? :D

File: lib/plausible/auth/totp.ex [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+  ## Management
+
+  State of TOTP for a particular user can be chcecked by calling `enabled?/1`.
+
+  TOTP can be disabled with `disable/2`. User is expected to provide their
+  current password for safety. Once disabled, all TOTP user settings are
+  cleared and any remaining generated recovery codes are removed. The function
+  can be safely run more than once.
+
+  If user needs to regenerate the recovery codes outside of setup procedure,
```
Comment (aerosol) [inline lib/plausible/auth/totp.ex:40]:
```suggestion
  If the user needs to regenerate the recovery codes outside of setup procedure,
```

File: lib/plausible/auth/totp.ex (hunk 1 above)
Comment (zoldar) [reply lib/plausible/auth/totp.ex:12]: No, safe, damn it. Thanks

File: lib/plausible/auth/totp.ex [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+      from(u in Plausible.Auth.User, where: u.id == ^user.id, select: u.totp_last_used_at)
+      |> Repo.one()
+
+    case datetime do
+      nil ->
+        nil
+
+      datetime ->
+        Timex.to_unix(datetime)
+    end
```
Comment (aerosol) [inline lib/plausible/auth/totp.ex:276]:
```suggestion
    if datetime do
      Timex.to_unix(datetime)
    end
```

File: lib/plausible/auth/totp/recovery_code.ex [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  @spec changeset_to_map(Ecto.Changeset.t(), NaiveDateTime.t()) :: map()
+  def changeset_to_map(changeset, now) do
+    changeset
+    |> apply_changes()
+    |> Map.take([:user_id, :code_digest])
+    |> Map.put(:inserted_at, now)
+  end
+
+  defp disambiguate(code) do
+    String.replace(code, "O", "8")
```
Comment (aerosol) [inline lib/plausible/auth/totp/recovery_code.ex:58]: Should we do the same for "l" ?

**Review by aerosol — APPROVED** [review]
Clean! ✨

File: lib/plausible/auth/totp/recovery_code.ex (hunk 5 above)
Comment (zoldar) [reply lib/plausible/auth/totp/recovery_code.ex]: https://github.com/plausible/analytics/pull/3525/commits/b3d511e5e99a88f33da89d46362806060daf6110

**Pull request merged**